### PR TITLE
[Hub Generated] Review request for Microsoft.MachineLearningServices to update version preview/2020-09-01-preview

### PR DESCRIPTION
**Pull request by forteddyt**
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/forteddyt/azure-rest-api-specs/thtodoro%2Fadd-mfe...master/).

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [ ] new API version 
      - [x] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “**WaitForARMFeedback**” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [ ] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://msazure.visualstudio.com/One/_workitems/create/Scenario?templateId=d1c215ec-5d6a-4cb3-a193-95ddd0a1ba96&ownerId=85bb92fb-21fe-41ca-9309-47f2c5f57515). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 1]
```diff
@@ -0,0 +1,6276 @@
...
+          "type": "string",
+          "readOnly": true
+        },
+        "properties": {
+          "description": "Additional attributes of the entity.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/CodeContainer"
+            }
+          ],
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:3126]: Anonymous `allOf` is not recommended
```suggestion
              "$ref": "#/definitions/CodeContainer",
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 2]
```diff
@@ -0,0 +1,6276 @@
...
+          "x-ms-client-flatten": true
+        },
+        "systemData": {
+          "description": "System data associated with resource provider",
+          "allOf": [
+            {
+              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData"
+            }
+          ],
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:3135]:
```suggestion
              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData",
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 3]
```diff
@@ -0,0 +1,6276 @@
...
+      "properties": {
+        "code": {
+          "type": "string"
+        },
+        "innerError": {
+          "allOf": [
+            {
+              "$ref": "#/definitions/InnerErrorResponse"
+            }
+          ]
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:3153]:
```suggestion
              "$ref": "#/definitions/InnerErrorResponse"
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 4]
```diff
@@ -0,0 +1,6276 @@
...
+          "items": {
+            "$ref": "#/definitions/RootError"
+          }
+        },
+        "innerError": {
+          "allOf": [
+            {
+              "$ref": "#/definitions/InnerErrorResponse"
+            }
+          ]
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:3201]:
```suggestion
              "$ref": "#/definitions/InnerErrorResponse"
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 5]
```diff
@@ -0,0 +1,6276 @@
...
+          "description": "The asset datastoreId",
+          "type": "string"
+        },
+        "assetPath": {
+          "description": "Details of asset path in the datastore.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/AssetPath"
+            }
+          ]
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:3256]:
```suggestion
              "$ref": "#/definitions/AssetPath"
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 6]
```diff
@@ -0,0 +1,6276 @@
...
+          "type": "string",
+          "readOnly": true
+        },
+        "properties": {
+          "description": "Additional attributes of the entity.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/CodeVersion"
+            }
+          ],
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:3323]:
```suggestion
              "$ref": "#/definitions/CodeVersion",
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 7]
```diff
@@ -0,0 +1,6276 @@
...
+          "x-ms-client-flatten": true
+        },
+        "systemData": {
+          "description": "System data associated with resource provider",
+          "allOf": [
+            {
+              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData"
+            }
+          ],
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:3332]:
```suggestion
              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData",
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 8]
```diff
@@ -0,0 +1,6276 @@
...
+          "type": "string",
+          "readOnly": true
+        },
+        "properties": {
+          "description": "Additional attributes of the entity.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/DataContainer"
+            }
+          ],
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:3405]:
```suggestion
              "$ref": "#/definitions/DataContainer",
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 9]
```diff
@@ -0,0 +1,6276 @@
...
+          "x-ms-client-flatten": true
+        },
+        "systemData": {
+          "description": "System data associated with resource provider",
+          "allOf": [
+            {
+              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData"
+            }
+          ],
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:3414]:
```suggestion
              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData",
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 10]
```diff
@@ -0,0 +1,6276 @@
...
+            "modelAsString": true
+          }
+        },
+        "accountKey": {
+          "description": "Storage account key authentication.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/AccountKeySection"
+            }
+          ]
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:3599]:
```suggestion
              "$ref": "#/definitions/AccountKeySection"
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 11]
```diff
@@ -0,0 +1,6276 @@
...
+        },
+        "certificate": {
+          "description": "Service principal certificate authentication.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/CertificateSection"
+            }
+          ]
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:3607]:
```suggestion
              "$ref": "#/definitions/CertificateSection"
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 12]
```diff
@@ -0,0 +1,6276 @@
...
+        },
+        "sas": {
+          "description": "Storage container SAS token authentication.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/SasSection"
+            }
+          ]
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:3615]:
```suggestion
              "$ref": "#/definitions/SasSection"
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 13]
```diff
@@ -0,0 +1,6276 @@
...
+        },
+        "servicePrincipal": {
+          "description": "Service principal password authentication.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/ServicePrincipalSection"
+            }
+          ]
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:3623]:
```suggestion
              "$ref": "#/definitions/ServicePrincipalSection"
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 14]
```diff
@@ -0,0 +1,6276 @@
...
+        },
+        "sqlAdmin": {
+          "description": "SQL user/password authentication.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/SqlAdminSection"
+            }
+          ]
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:3631]:
```suggestion
              "$ref": "#/definitions/SqlAdminSection"
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 15]
```diff
@@ -0,0 +1,6276 @@
...
+      ],
+      "type": "object",
+      "properties": {
+        "credentials": {
+          "description": "Azure Data Lake credentials.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/DatastoreCredentials"
+            }
+          ]
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:3649]:
```suggestion
              "$ref": "#/definitions/DatastoreCredentials"
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 16]
```diff
@@ -0,0 +1,6276 @@
...
+      ],
+      "type": "object",
+      "properties": {
+        "credentials": {
+          "description": "Azure SQL database credentials.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/DatastoreCredentials"
+            }
+          ]
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:3675]:
```suggestion
              "$ref": "#/definitions/DatastoreCredentials"
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 17]
```diff
@@ -0,0 +1,6276 @@
...
+            "read"
+          ]
+        },
+        "credentials": {
+          "description": "Azure SQL database credentials.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/DatastoreCredentials"
+            }
+          ]
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:3741]:
```suggestion
              "$ref": "#/definitions/DatastoreCredentials"
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 18]
```diff
@@ -0,0 +1,6276 @@
...
+      ],
+      "type": "object",
+      "properties": {
+        "credentials": {
+          "description": "Azure SQL database credentials.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/DatastoreCredentials"
+            }
+          ]
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:3799]:
```suggestion
              "$ref": "#/definitions/DatastoreCredentials"
```

**Review by ArcturusZhang — CHANGES_REQUESTED** [review]
We have a lot of anonymous allOf in this swagger, which is highly not recommended.
This is the first round of review... to be continued...

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 19]
```diff
@@ -0,0 +1,6276 @@
...
+              "readOnly": true
+            },
+            "containerName": {
+              "description": "The container name to which the labels will be exported.",
+              "type": "string",
+              "readOnly": true
+            }
+          }
+        }
+      ],
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:5996]:
```suggestion
      "allOf": [
        {
          "$ref": "#/definitions/ExportSummary"
        }
        ],
          "type": "object",
          "properties": {
            "snapshotPath": {
              "description": "The output path where the labels will be exported.",
              "type": "string",
              "readOnly": true
            },
            "containerName": {
              "description": "The container name to which the labels will be exported.",
              "type": "string",
              "readOnly": true
            }
          },
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 20]
```diff
@@ -0,0 +1,6276 @@
...
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:6020]:
```suggestion
      "allOf": [
        {
          "$ref": "#/definitions/ExportSummary"
        }
        ],
          "type": "object",
          "properties": {
            "snapshotPath": {
              "description": "The output path where the labels will be exported.",
              "type": "string",
              "readOnly": true
            },
            "containerName": {
              "description": "The container name to which the labels will be exported.",
              "type": "string",
              "readOnly": true
            }
          },
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 21]
```diff
@@ -0,0 +1,6276 @@
...
+          "type": "object",
+          "properties": {
+            "labeledAssetName": {
+              "description": "The unique name of the labeled data asset.",
+              "type": "string",
+              "readOnly": true
+            }
+          }
+        }
+      ],
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:6039]:
```suggestion
      "allOf": [
        {
          "$ref": "#/definitions/ExportSummary"
        }
        ],
          "type": "object",
          "properties": {
            "labeledAssetName": {
              "description": "The unique name of the labeled data asset.",
              "type": "string",
              "readOnly": true
            }
          },
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 22]
```diff
@@ -0,0 +1,6276 @@
...
+          "type": "string",
+          "readOnly": true
+        },
+        "properties": {
+          "description": "Additional attributes of the entity.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/ModelContainer"
+            }
+          ],
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:6103]:
```suggestion
              "$ref": "#/definitions/ModelContainer",
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 23]
```diff
@@ -0,0 +1,6276 @@
...
+          "x-ms-client-flatten": true
+        },
+        "systemData": {
+          "description": "System data associated with resource provider",
+          "allOf": [
+            {
+              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData"
+            }
+          ],
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:6112]:
```suggestion
              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData",
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 24]
```diff
@@ -0,0 +1,6276 @@
...
+          "description": "The asset datastoreId",
+          "type": "string"
+        },
+        "assetPath": {
+          "description": "Details of asset path in the datastore.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/AssetPath"
+            }
+          ]
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:6164]:
```suggestion
              "$ref": "#/definitions/AssetPath"
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 25]
```diff
@@ -0,0 +1,6276 @@
...
+          "type": "string",
+          "readOnly": true
+        },
+        "properties": {
+          "description": "Additional attributes of the entity.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/ModelVersion"
+            }
+          ],
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:6231]:
```suggestion
              "$ref": "#/definitions/ModelVersion",
```

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 26]
```diff
@@ -0,0 +1,6276 @@
...
+          "x-ms-client-flatten": true
+        },
+        "systemData": {
+          "description": "System data associated with resource provider",
+          "allOf": [
+            {
+              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData"
+            }
+          ],
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:6240]:
```suggestion
              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData",
```

**Review by ArcturusZhang — CHANGES_REQUESTED** [review]
Second part of anonymous `allOf` issues...

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json (hunk 1 above)
Comment (forteddyt) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:3126]: OpenApi 2.0 spec specifies that sibling properties will be ignored if `$ref` is present - https://github.com/OAI/OpenAPI-Specification/issues/1514. We need `"x-ms-client-flatten": true` and (to a lesser extent) the description properties, so we encompass all $references with AllOf

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 27]
```diff
@@ -0,0 +1,6276 @@
...
+        {
+          "type": "object",
+          "properties": {
+            "dockerImageUri": {
+              "description": "Image name of a custom base image.\r\n<seealso href=\"https://docs.microsoft.com/en-us/azure/machine-learning/how-to-deploy-custom-docker-image#use-a-custom-base-image\" />",
+              "type": "string"
+            }
+          }
+        }
+      ],
```
[the remark this answers is not in the document]
Comment (forteddyt) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:4266]: Can I get an explanation on why this is preferred? AllOf is supposed to take in an [array of object definitions](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#composition-and-inheritance-polymorphism) that compose a single object. This is the intention of what we are doing with our AllOfs

**Comment by forteddyt** [comment]
Ready for re-review, @ArcturusZhang. Also could you please address my questions?

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json (hunk 1 above)
Comment (ArcturusZhang) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json]: Despite the OpenApi has this restriction, but autorest treats this case differently. Autorest could properly identify the description here. @PhoenixHe-msft to confirm

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json (hunk 27 above)
[the remark this answers is not in the document]
Comment (forteddyt) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json]: @ArcturusZhang

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 28]
```diff
@@ -0,0 +1,6276 @@
...
+            "trialComponent": {
+              "allOf": [
+                {
+                  "$ref": "#/definitions/TrialComponent"
+                }
+              ]
+            }
+          }
+        }
+      ],
```
[the remark this answers is not in the document]
Comment (forteddyt) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json]: you're right, but I will ignore missing descriptions for now

**Review by ArcturusZhang — APPROVED** [review]

**Comment by forteddyt** [comment]
All future new specs will reference common types. Reverting references on this 2020-09-01-preview

**Comment by ArcturusZhang** [comment]
Hey, do we have any update on this PR?

**Comment by forteddyt** [comment]
hey, sorry for the delay. Yes, in addition to updating the spec with the latest (minor) changes, we added AutoMlJob and PipelineJob to the existing Job API. We also added a new Component API, that has ComponentContainer and ComponentVersion.

Could these new additions be reviewed?

**Comment by forteddyt** [comment]
I'm (expectedly) getting the same LintDiff issues in this PR as in my [2021-03-01-preview PR](https://github.com/Azure/azure-rest-api-specs/pull/12554), so I shall voice my questions/concerns here:

1. Regarding [R2008](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/openapi-authoring-automated-guidelines.md#R2008), I was talking with another ARM reviewer last week and they were saying the issue was that the `$ref` definition itself wasn't `readOnly: true`, but the definition that used the reference was `readOnly: true`. To that, I'm wondering: why is that an issue? What if a `$ref` definition is used as `readOnly: true` in one definition, but is mutable in another? for example: 
```json
"output": {
   "ref" :"#/definitions/Job",
   "readOnly": true,
   "x-ms-mutability": ["read"]
}
"input": {
   "ref": "#/definitions/Job",
   "readOnly": false,
   "x-ms-mutability": ["read", "write"]
}
```
2. Regarding [R2026](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/openapi-authoring-automated-guidelines.md#R2026), in our codebase this is represented as a `Dictionary<string, Dictionary<string, string>>`. If the current definition is truly incorrect, I'm having trouble understanding how to properly represent this in the spec... Could I get some guidance on this?
3. Regarding [R3015](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/openapi-authoring-automated-guidelines.md#R3015), I was told by a previous ARM reviewer to ignore this. "enum" is the name of the property, not an Enum-type property

**Comment by ArcturusZhang** [comment]
CC @jianyexi for the Linter questions

**Comment by forteddyt** [comment]
Additionally, we added the Schedule API

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json [hunk 29]
```diff
@@ -9788,6 +10010,301 @@
...
+            "Enabled",
+            "Disabled"
+          ],
+          "type": "string",
+          "x-ms-enum": {
+            "name": "ScheduleStatus",
+            "modelAsString": true
+          }
+        },
+        "provisioningStatus": {
```
Comment (forteddyt) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:10149]: This is **not** meant to be the `provisioningState` property used in asynchronous operations, as defined [here](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#provisioningstate-property). Will the name similarity be an issue?

**Comment by pilor** [comment]
>         "createdTimeUtc": {

This should be in systemData.  You shouldn't duplicate it inside the resource properties

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:8568 in 9f78980. [](commit_id = 9f789805b324a5ab3b788475f6ae72ed0f55ce8e, deletion_comment = False)

**Comment by pilor** [comment]
>     "ResourceIdentityInArm": {

please mark all managed identity properties readonly that are readonly

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:9097 in 9f78980. [](commit_id = 9f789805b324a5ab3b788475f6ae72ed0f55ce8e, deletion_comment = False)

**Comment by pilor** [comment]
>         "clientSecretUrl": {

this is not something you expose to the user.  Please read through the managed identity implementation/contract docs one more time and ensure your contract is correct

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:9085 in 9f78980. [](commit_id = 9f789805b324a5ab3b788475f6ae72ed0f55ce8e, deletion_comment = False)

**Comment by pilor** [comment]
>         "type": {

Don't use "type" as the discriminator on all these definitions.  Change it to something more specific like "scheduleType" for schedules, etc...  You got exceptions to use "tags", etc... within the property bag but there is no reason to duplicate all top level property names inside your property bag when you can easily use something else

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:10184 in 9f78980. [](commit_id = 9f789805b324a5ab3b788475f6ae72ed0f55ce8e, deletion_comment = False)

**Comment by forteddyt** [comment]
> > ```
> > "ResourceIdentityInArm": {
> > ```
> 
> please mark all managed identity properties readonly that are readonly
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:9097 in [9f78980](https://github.com/Azure/azure-rest-api-specs/commit/9f789805b324a5ab3b788475f6ae72ed0f55ce8e). [](commit_id = [9f78980](https://github.com/Azure/azure-rest-api-specs/commit/9f789805b324a5ab3b788475f6ae72ed0f55ce8e), deletion_comment = False)

From my understanding, ManagedIdentities are updatable. Our service supports updating them, as well

**Comment by pilor** [comment]
>         "identityType": {

"type" was the correct property name

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:8857 in 3c977ff. [](commit_id = 3c977ff0379a9c1251401f977b78fd2f715eb728, deletion_comment = False)

**Comment by pilor** [comment]
>     "ResourceIdentityInArm": {

"principalId" and "tenantId" still need to be marked readonly

---
In reply to: [808519549](https://github.com/Azure/azure-rest-api-specs/pull/12508#issuecomment-808519549) [](ancestors = 808519549)

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:9097 in 9f78980. [](commit_id = 9f789805b324a5ab3b788475f6ae72ed0f55ce8e, deletion_comment = False)

**Comment by pilor** [comment]
>           "type": "string"

"resourceId" and "tenantId" are not part of the contract.  See [https://armwiki.azurewebsites.net/authorization/managed_identities/MSIOnboardingUserAssigned.html#resource-definition-response](https://armwiki.azurewebsites.net/authorization/managed_identities/MSIOnboardingUserAssigned.html#resource-definition-response)

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:8849 in 3c977ff. [](commit_id = 3c977ff0379a9c1251401f977b78fd2f715eb728, deletion_comment = False)

**Comment by forteddyt** [comment]
> > ```
> >       "type": "string"
> > ```
> 
> "resourceId" and "tenantId" are not part of the contract. See https://armwiki.azurewebsites.net/authorization/managed_identities/MSIOnboardingUserAssigned.html#resource-definition-response
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:8849 in [3c977ff](https://github.com/Azure/azure-rest-api-specs/commit/3c977ff0379a9c1251401f977b78fd2f715eb728). [](commit_id = [3c977ff](https://github.com/Azure/azure-rest-api-specs/commit/3c977ff0379a9c1251401f977b78fd2f715eb728), deletion_comment = False)

I've updated ResourceIdentityInArm's definition (alongside renaming it to ResourceIdentity) in c5e2292. Could you take a look and tell me if it's properly defined? I'll update the example files afterwards

File: specification/machinelearningservices/resource-manager/readme.md [hunk 30]
```diff
@@ -25,6 +25,10 @@ To see additional help and options, run:
 These are the global settings for the Machine Learning Services API.
 
 ``` yaml
+<<<<<<< HEAD
+openapi-type: arm 
+tag: package-2020-09-01-preview
+=======
```
Comment (ArcturusZhang) [inline specification/machinelearningservices/resource-manager/readme.md:31]: Hey, looks like that we have some conflicts that are not correctly resolved here

**Pull request closed without merging**